README.md
# NAME

API::Chaplus - It's an API module for [Chaplus](https://www.chaplus.jp/) with perl

# SYNOPSIS

    use API::Chaplus;
    my $api    = API::Chaplus->new( apikey => 'your_key' );
    my $answer = $api->bestResponse( utterance => '調子どう？' );
    print $answer;    # お返事が表示される

# DESCRIPTION

API::Chaplus is a test implement for [Chaplus-API](https://k-masashi.github.io/chaplus-api-doc/)

## CONSTRUCTOR

### new( apikey => YOUR\_KEY )

attribute 'apikey' is required

## METHODS

### request( utterance => '日本語で話しかけよう', ... )

attribute 'utterance' is required

returns Perl hash decoded from JSON 

### bestResponse( utterance => '日本語で話しかけよう', ... )

attribute 'utterance' is required

returns Response object

### responses( utterance => '日本語で話しかけよう', ... )

attribute 'utterance' is required

returns list of response objects

### tokenized( utterance => '日本語で話しかけよう', ... )

attribute 'utterance' is required

returns list of tokenized objects

### options( utterance => '日本語で話しかけよう', ... )

attribute 'utterance' is required

returns list of options objects

# SEE ALSO

- [Botが雑談に応えられるようにするWEB APIを公開してみた話【個人開発】](https://qiita.com/maKunugi/items/b1afb6441571119729a7)
- [chaplus-api-doc](https://k-masashi.github.io/chaplus-api-doc/)
- [API::Chaplus::Request.md](https://github.com/worthmine/API-Chaplus/blob/master/lib/API/Chaplus/Request.md)

# LICENSE

Copyright (C) worthmine.

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

# AUTHOR

worthmine <worthmine@gmail.com>


Changes
Revision history for Perl extension API-Chaplus

{{$NEXT}}

    - original version



lib/API/Chaplus/Request.pm
package API::Chaplus::Request;
use 5.008001;
use Carp;

our $VERSION = "0.01";

use Mouse;
use Mouse::Util::TypeConstraints;

use API::Chaplus::Request::AgentState;
use API::Chaplus::Request::Addition;
use URL::Encode qw(url_encode_utf8 url_decode_utf8);

use overload (
    '""'  => sub { url_decode_utf8 $_[0]->utterance() },
    '<=>' => sub { $_[0]->score <=> $_[1] },
);

has username => ( is => 'rw', isa => 'Encoded', coerce => 1 );

subtype 'Utterance' => as 'Encoded' =>
  where { /^[A-F0-9%]+$/ and length($_) > 0 }
=> message { "empty utterance was set" };
coerce 'Utterance' => from 'Str' => via { url_encode_utf8($_) };
has utterance => ( is => 'rw', isa => 'Utterance', required => 1, coerce => 1 );

subtype 'AgentState' => as 'HashRef';
coerce 'AgentState', from 'API::Chaplus::Request::AgentState',
  via { API::Chaplus::Request::Addition::serialize($_) };
has agentState => ( is => 'rw', isa => 'AgentState', coerce => 1 );

subtype 'Addition' => as 'HashRef';
coerce 'Addition', from 'API::Chaplus::Request::Addition', via {
    return +{%$_}
};
has addition => ( is => 'rw', isa => 'Addition', coerce => 1 );

no Mouse;

1;
__END__

=encoding utf-8

=head1 NAME

API::Chaplus::Request - objectifier for Requests 

=head1 SYNOPSIS

    use API::Chaplus::Request;
    my $pair = API::Chaplus::UtterancePair->new(
        utterance => '肩凝った', 
        response => '適度に運動しないとね',
    );

    my $addition = API::Chaplus::Request::Addition->new(       
        options        => [qw(疲れた 肩凝った)],
        utterancePairs => $pair,                    # auto optimize to [$pair] by Mouse Coercion
        ngwords        => [ "ため息", "やめてしまえ" ],
    );
    
    my $state = API::Chaplus::Request::AgentState->new(
        agentName => 'エージェント',
        tone      => 'kansai',
        age       => '14歳',
    );

    my $req =  use API::Chaplus::Request->new(
        username   => '太郎',
        utterance  => '調子はどう？',
        agentState => $state,
        addition   => $addition,
    );
    my $api    = API::Chaplus->new( apikey => 'your_key' );
    my $answer = $api->bestResponse($req);
    print $answer;    # お返事が表示される

=head1 DESCRIPTION

API::Chaplus::Request provides you objectified attributes for API::Chaplus; 

=head2 CONSTRUCTOR

=head3 new()

Attribute 'utterance' is required. 
Any other attributes are optional

=head2 METHODS

=head3 username(I<name>)
accepts/returns the name 

=head3 utterance(I<utterance>)

accepts/returns the utterance 

=head3 agentState(I<agentState object>)

accepts/returns 'agentState' Object 

=head3 addition(I<addition object>)

accepts/returns 'addition' Object 

=head1 SEE ALSO

=over

=item L<chaplus-api-doc|https://k-masashi.github.io/chaplus-api-doc/>

=item L<API::Chaplus|https://github.com/worthmine/API-Chaplus>

=back

=head1 LICENSE

Copyright (C) worthmine.

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=head1 AUTHOR

worthmine E<lt>worthmine@gmail.comE<gt>

=cut




lib/API/Chaplus/Request/Addition.pm
package API::Chaplus::Request::Addition;
use 5.008001;
use Carp;

our $VERSION = "0.01";

use Mouse;
use Mouse::Util::TypeConstraints;

use API::Chaplus::UtterancePair;
use URL::Encode qw(url_encode_utf8);

subtype 'NotEncoded' => as 'Str' => where { $_ !~ /^[A-F0-9%]+$/ };
subtype 'Encoded'    => as 'Str' => where { /^[A-F0-9%]+$/ };
coerce 'Encoded' => from 'NotEncoded' => via { url_encode_utf8($_) };

subtype 'Options' => as 'ArrayRef[Encoded]';
coerce 'Options', from 'ArrayRef[NotEncoded]', via {
    [ map { url_encode_utf8($_) } @$_ ]
}, from 'NotEncoded', via { [ url_encode_utf8($_) ] };
has options => ( is => 'rw', isa => 'Options', coerce => 1 );

subtype 'UtterancePairs' => as 'ArrayRef[HashRef]';
coerce 'UtterancePairs', from 'ArrayRef[API::Chaplus::UtterancePair]', via {
    [ map { serialize($_) } @$_ ];
}, from 'API::Chaplus::UtterancePair', via { [ serialize($_) ] };
has utterancePairs => ( is => 'rw', isa => 'UtterancePairs', coerce => 1 );

subtype 'Ngwords' => as 'ArrayRef[Encoded]';
coerce 'Ngwords', from 'ArrayRef[NotEncoded]', via {
    return [ map { url_encode_utf8($_) } @$_ ]
}, from 'NotEncoded', via { [ url_encode_utf8($_) ] };
has ngwords => ( is => 'rw', isa => 'Ngwords', coerce => 1 );

no Mouse;

sub serialize {
    my $obj   = shift;
    my $pairs = {};
    while ( my ( $key, $value ) = each %$obj ) {
        $pairs->{$key} = url_encode_utf8($value);
    }
    return $pairs;
}

1;
__END__





lib/API/Chaplus.pm
package API::Chaplus;
use 5.008001;
use strict;
use warnings;
use Carp;

use Data::Dumper::AutoEncode qw(eDumper);
use JSON;
my $j = JSON->new()->utf8();

use URL::Encode qw(url_encode_utf8);
use Furl;
use HTTP::Request::JSON;
my $rj = HTTP::Request::JSON->new();

use API::Chaplus::Response;
use API::Chaplus::Tokenized;
use API::Chaplus::Options;

our $VERSION = "0.01";

=encoding utf-8

=head1 NAME

API::Chaplus - It's an API module for L<Chaplus|https://www.chaplus.jp/> with perl

=head1 SYNOPSIS

    use API::Chaplus;
    my $api    = API::Chaplus->new( apikey => 'your_key' );
    my $answer = $api->bestResponse( utterance => '調子どう？' );
    print $answer;    # お返事が表示される

=head1 DESCRIPTION

API::Chaplus is a test implement for L<Chaplus-API|https://k-masashi.github.io/chaplus-api-doc/>

=head2 CONSTRUCTOR

=head3 new( apikey => YOUR_KEY )

attribute 'apikey' is required

=cut

sub new {
    my $class = shift;
    my $self  = bless {}, $class;
    my %attr  = @_;

    $self->{'apikey'} = $attr{'apikey'} || croak "missing param 'apikey'";
    $self->{'endPoint'} =
      'https://www.chaplus.jp/v1/chat?apikey=' . $attr{'apikey'};

    return $self;
}

=head2 METHODS

=head3 request( utterance => '日本語で話しかけよう', ... )

attribute 'utterance' is required

returns Perl hash decoded from JSON 

=cut 

sub request {
    my $self = shift;
    my $req  = undef;
    my $res  = undef;
    my %attr = ();
    if ( @_ > 1 ) {
        %attr = map { url_encode_utf8 $_ } @_;
        croak "missing key utterance" unless $attr{'utterance'};
    }
    elsif ( ref $_[0] eq 'API::Chaplus::Request' ) {
        my $req = shift;
        %attr = %$req;
    }

    #::note ::explain \%attr;
    my $params = $rj->json_content( \%attr );
    my $f      = Furl->new;
    unless ( defined $req ) {
        $req = Furl::Request->new( 'POST', $self->{'endPoint'},
            { 'Content-Type' => 'application/json' }, $params );

        ($res) = $f->request($req);
    }

    if ( exists $res->{'code'} and $res->{'code'} eq '200' ) {
        return $j->decode( $res->{'content'} );
    }
    else {
        croak "something wrong to post by Furl: " . eDumper($res);
    }
}

=head3 bestResponse( utterance => '日本語で話しかけよう', ... )

attribute 'utterance' is required

returns Response object

=cut 

sub bestResponse {
    my $self = shift;
    my $q    = $self->request(@_);
    return API::Chaplus::Response->new( $q->{'bestResponse'} );
}

=head3 responses( utterance => '日本語で話しかけよう', ... )

attribute 'utterance' is required

returns list of response objects

=cut 

sub responses {
    my $self = shift;
    my $q    = $self->request(@_);
    my @list = map { API::Chaplus::Response->new($_) } @{ $q->{'responses'} };
    return wantarray ? @list : \@list;
}

=head3 tokenized( utterance => '日本語で話しかけよう', ... )

attribute 'utterance' is required

returns list of tokenized objects

=cut 

sub tokenized {
    my $self = shift;
    my $q    = $self->request(@_);
    my @list = map { API::Chaplus::Tokenized->new( token => $_ ) }
      @{ $q->{'tokenized'} };
    return wantarray ? @list : \@list;
}

=head3 options( utterance => '日本語で話しかけよう', ... )

attribute 'utterance' is required

returns list of options objects

=cut 

sub options {
    my $self = shift;
    my $q    = $self->request(@_);
    my @list =
      map { API::Chaplus::Options->new( suggest => $_ ) } @{ $q->{'options'} };
    return wantarray ? @list : \@list;
}

1;
__END__

=head1 SEE ALSO

=over

=item L<Botが雑談に応えられるようにするWEB APIを公開してみた話【個人開発】|https://qiita.com/maKunugi/items/b1afb6441571119729a7>

=item L<chaplus-api-doc|https://k-masashi.github.io/chaplus-api-doc/>

=item L<API::Chaplus::Request.md|https://github.com/worthmine/API-Chaplus/blob/master/lib/API/Chaplus/Request.md>

=back

=head1 LICENSE

Copyright (C) worthmine.

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=head1 AUTHOR

worthmine E<lt>worthmine@gmail.comE<gt>

=cut
